src/LibraBFT/Impl/Consensus/BlockStorage/BlockTree.agda

```diff
@@ -262,6 +263,50 @@ abstract
   insertQuorumCertE-Either-≡ : insertQuorumCertE-Either ≡ insertQuorumCertE.E
   insertQuorumCertE-Either-≡ = refl
 
+module insertQuorumCertE-AST (qc : QuorumCert) (bt0 : BlockTree) where
+  open EitherAST ErrLog
+  open SyntaxExt renaming (bail to bail-AST; return to return-AST)
+  open BranchingSyntax EitherOps
+
+  here' : List String → List String
+  here' t = "BlockTree" ∷ "insertQuorumCert" ∷ t
+
+  blockId = qc ^∙ qcCertifiedBlock ∙ biId
+
+  safetyInvariant = forM_ (Map.elems (bt0 ^∙ btIdToQuorumCert)) $ \x →
+          lcheck (   (x  ^∙ qcLedgerInfo ∙ liwsLedgerInfo ∙ liConsensusDataHash
+                  ==  qc ^∙ qcLedgerInfo ∙ liwsLedgerInfo ∙ liConsensusDataHash)
+                  ∨  (x  ^∙ qcCertifiedBlock ∙ biRound
+                  /=  qc ^∙ qcCertifiedBlock ∙ biRound))
+                 (here' ("failed check" ∷ "existing qc == qc || existing qc.round /= qc.round" ∷ []))
+
+  continue1 : BlockTree → HashValue → ExecutedBlock → List InfoLog → (BlockTree × List InfoLog)
+  continue2 : BlockTree → List InfoLog → (BlockTree × List InfoLog)
+
+  insertQuorumCertE-AST : EitherDExt (BlockTree × List InfoLog)
+  insertQuorumCertE-AST =
+    case safetyInvariant of λ where
+      (Left  e)    → bail-AST e
+      (Right unit) → maybeSD (btGetBlock blockId bt0) (bail-AST fakeErr) λ block →
+                     maybeSD (bt0 ^∙ btHighestCertifiedBlock) (bail-AST fakeErr) λ hcb →
+                     ifAST ⌊ (block ^∙ ebRound) >? (hcb ^∙ ebRound) ⌋
+                     then
+                       (let bt   = bt0 & btHighestCertifiedBlockId ∙~ block ^∙ ebId
+                                       & btHighestQuorumCert       ∙~ qc
+                            info = (fakeInfo ∷ [])
+                        in return-AST (continue1 bt  blockId block info))
+                     else
+                       (return-AST (continue1 bt0 blockId block []))
+
+  continue1 bt blockId block info =
+    continue2 ( bt & btIdToQuorumCert ∙~ lookupOrInsert blockId qc (bt ^∙ btIdToQuorumCert))
+              ( (fakeInfo ∷ info) ++ (if ExecutedBlock.isNilBlock block then fakeInfo ∷ [] else [] ))
```

Comment:
I assume you mean to put `fakeInfo :: []` instead of the expression on line 303?  That breaks the proof and I don't see the benefit of changing to something less faithful to the Haskell code, just to be consistent with with other (alleged) places that are less faithful.   NO CHANGE.